src/shared_realm.cpp

```diff
@@ -181,6 +191,24 @@ void Realm::open_with_config(const Config& config,
         util::File::remove(config.path);
         open_with_config(config, history, shared_group, read_only_group, realm);
     }
+#if REALM_ENABLE_SYNC
+    catch (IncompatibleHistories const& ex) {
+        if (!server_synchronization_mode || !is_nonupgradable_history(ex))
+            translate_file_exception(config.path, config.immutable()); // Throws
+
+        // Move the Realm file into the recovery directory.
+        std::string recovery_directory = SyncManager::shared().recovery_directory_path();
+        std::string new_realm_path = util::reserve_unique_file_name(recovery_directory, "synced-realm-XXXXXXX");
+        util::File::move(config.path, new_realm_path);
+
+        std::string message = util::format("The local copy of this synced Realm was created with an incompatible "
```

Comment:
This is worded as a message a user writing code in an interactive environment. It doesn't really work as either an error displayed to the user in a app, or to a developer in a crash log after they failed to handle it.

Reply:
The nature of this error is that the developer needs to be proactively accommodate it when they update their app from sync v1.x to 2.x. It’s not intended to be displayed to users, and if it crashes their app in production its too late for them to do anything about it.

Reply:
Yes, that's my point. It's worded as a message to an end-user rather than something telling a developer that they need to write code to handle this case. Things like "you can now retry doing X" make sense when you're telling to user to click on the button they just clicked on again, but less so when they actually need to write code to handle this error case.

Reply:
@tgoyne, would you mind suggesting an edit to this error message? I'm struggling to come up with something that's an improvement, and this is the only part of your feedback that I've not yet addressed. I'd like to get this merged ASAP.

Reply:
I was thinking something like this:

```
That the local copy of this synced Realm was created with an incompatible version of Realm. The existing file has been moved to '%1'. The application developer is responsible for catching this error, opening the backup copy of the Realm to recover any pending changes, and opening the synchronized Realm again.
```

Reply:
"The local copy of this sync Realm was created with an incompatible version of Realm. It has been moved aside, and the Realm will be re-downloaded the next time it is opened. You will need to write a handler for this error which either uses the attached configuration to open the Realm in read-only mode to recover any pending changes or deletes the old Realm file. See http://realm.io/???? for more information about this migration."

I'm not terribly happy with this wording either, but the idea behind it is that the message is trying to tell the developer what code they need to write. I'm assuming that we will have some sort of news post about this which should be linked. The path is not included in the error because it's hard to imagine a situation where it'd be useful, while it's easy to imagine situations where people do bad things with it (like trying to extract the path from the error message rather than getting it from the config) because it's the most obvious approach.

Also now I'm wondering if we should do something more like how migration blocks work since writing code which handles a Realm instance failing to open, doing stuff, and then retrying creating the Realm instance is probably going to be really awkward for people that don't have a wrapper factory function for getting Realm instances, but maybe making that sufficiently flexible would get too complicated.

Reply:
I've updated to use your wording with a few minor tweaks. I've omitted the link since there's nothing to link to for now. Bindings can append their own at that level if they choose.

I think the number of people that'll need to deal with this is relatively small, so I'm ok with the API being a little clunky (particularly given the time we have available and the need for it to be implemented in each binding).